cdap-app-fabric/src/main/java/io/cdap/cdap/reporting/ProgramHeartbeatTable.java

```diff
@@ -121,6 +123,32 @@ public void deleteAll() throws IOException {
     return resultRunRecordList;
   }
 
+  /**
+   * Find al running pipelines as of the supplied timestamp, for each of the scanned rows, we maintain the latest {@link
+   * RunRecordDetail} identified by {@link ProgramRunId}, Since there can be more than one RunRecordDetail for the same
+   * runId due to multiple state changes and heart beat messages.
+   *
+   * @param namespaces set of namespaces
+   * @return collection of {@link RunRecordDetail}
+   */
+  public Collection<RunRecordDetail> findRunningAtTimestamp(long runningOnTimestamp,
+                                                            Set<String> namespaces) throws IOException {
+    List<RunRecordDetail> resultRunRecordList = new ArrayList<>();
+    for (String namespace : namespaces) {
```

Comment:
Generally I would against perform table query in a loop. You can perform one query by using the `StructuredTable.multiScan` method.

Reply:
I updated this class to use the multiScan method.